[ADD] mail_notification_custom_subject: Apply a custom subject to mail notifications

joao-p-marques opened the pull request:
@Tecnativa
TT27191

ping @pedrobaeza

victoralmau commented on mail_notification_custom_subject/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+    "installable": True,
+    "depends": [
+        "mail",
+        "account", # needed for tests
```
```suggestion
```
Need to user base models (`res.partner` model for example).

victoralmau commented on mail_notification_custom_subject/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+    "version": "12.0.1.0.0",
+    "category": "Social Network",
+    "website": "https://github.com/OCA/social/",
+    "author": "Tecnativa, Eficent, Odoo Community Association (OCA)",
```
```suggestion
    "author": "Tecnativa, Odoo Community Association (OCA)",
```

victoralmau commented on mail_notification_custom_subject/models/mail_thread.py:
```diff
@@ -0,0 +1,32 @@
+            mail_auto_delete=mail_auto_delete, **kwargs
+        )
+        mail_template = self.env["mail.message.custom.subject"].search([
+            ("model_id.model", "=", self._name),
```
```suggestion
            ("model_id.model", "=", vals['model']),
```

joao-p-marques replied:
I ended up not applying this because, after the latest re-factor, I don't have `vals` available, and `self._name` seems like the best option.

victoralmau commented on mail_notification_custom_subject/models/mail_thread.py:
```diff
@@ -0,0 +1,32 @@
+            ("model_id.model", "=", self._name),
+            ("subtype_ids", "=", vals["subtype_id"].id),
+        ])
+        subject_override = mail_template.render_subject_template(self._name, self.id)
```
```suggestion
        subject_override = mail_template.render_subject_template(vals['model'], vals['res_id'])
```

joao-p-marques replied:
Same here as above.

victoralmau commented on mail_notification_custom_subject/models/mail_message_custom_subject.py:
```diff
@@ -0,0 +1,158 @@
+        string='Subject Template', required=True,
+        help="Subject (placeholders may be used here)"
+    )
+    replace = fields.Boolean(
+        string="Replace", default=True,
+        help="Whether to replace or append to other templates that apply to a"
+        " given context"
+    )
```
Append allways after right?
Maybe need to change `replace` field to `position` (select): replace/append_after/append_before

What do you think?

joao-p-marques replied:
That may be an interesting option. @pedrobaeza WDYT?

pedrobaeza replied:
OK

victoralmau commented on mail_notification_custom_subject/tests/test_mail_notification_custom_subject.py:
```diff
@@ -0,0 +1,108 @@
+            'name': 'Test partner',
+            'supplier': True,
+        })
+        receivable = self.env.ref('account.data_account_type_receivable').id
+        expenses = self.env.ref('account.data_account_type_expenses').id
+        self.invoice_account = self.env['account.account'].search(
+            [('user_type_id', '=', receivable)], limit=1).id
+        self.invoice_line_account = self.env['account.account'].search(
+            [('user_type_id', '=', expenses)], limit=1).id
+        product = self.env['product.product'].create({
+            'name': 'Product test',
+            'type': 'service',
+        })
+        self.invoice_1 = self.env['account.invoice'].create({
+            'partner_id': self.partner.id,
+            'account_id': self.invoice_account,
+            'type': 'in_invoice',
+            'invoice_line_ids': [(0, 0, {
+                'product_id': product.id,
+                'quantity': 1.0,
+                'price_unit': 100.0,
+                'name': 'product that cost 100',
+                'account_id': self.invoice_line_account,
+            })]
+        })
+        self.invoice_2 = self.env['account.invoice'].create({
+            'partner_id': self.partner.id,
+            'account_id': self.invoice_account,
+            'type': 'in_invoice',
+            'invoice_line_ids': [(0, 0, {
+                'product_id': product.id,
+                'quantity': 1.0,
+                'price_unit': 100.0,
+                'name': 'product that cost 100',
+                'account_id': self.invoice_line_account,
+            })]
+        })
+        self.invoice_2.number = 'INV/TEST'
```
```suggestion
```
Need to remove everithing about `account` addon and use `res.partner` model

victoralmau commented on mail_notification_custom_subject/tests/test_mail_notification_custom_subject.py:
```diff
@@ -0,0 +1,108 @@
+        self.env["mail.message.custom.subject"].search([]).unlink()
+        self.template = self.env["mail.message.custom.subject"].create({
+            "name": "Test template 1",
+            "model_id": self.env.ref("account.model_account_invoice").id,
```
```suggestion
            "model_id": self.env.ref("base.model_res_partner").id,
```

victoralmau commented on mail_notification_custom_subject/tests/test_mail_notification_custom_subject.py:
```diff
@@ -0,0 +1,108 @@
+            "subtype_ids": [(6, 0, [
+                self.env.ref('mail.mt_comment').id,
+            ])],
+            "subject_template": "${object.number or 'n/a'} and something more",
```
```suggestion
            "subject_template": "${object.name or 'n/a'} and something more",
```

victoralmau commented on mail_notification_custom_subject/tests/test_mail_notification_custom_subject.py:
```diff
@@ -0,0 +1,108 @@
+        self.assertEquals(mail_message_2.subject, self.invoice_2.number + " and something more")
+
+        # Explicit subject should also be overwritten
+        mail_message_3 = self.invoice_2.message_post(body='Test', subtype='mail.mt_comment', subject="Test")
+        # Get message and check subject
+        self.assertEquals(mail_message_3.subject, self.invoice_2.number + " and something more")
+
+    def test_email_subject_template_invoice_normal(self):
+        # Send note in invoice
+        mail_message_1 = self.invoice_1.message_post(body='Test', subtype='mail.mt_note', subject="Test")
+        # Get message and check subject
+        self.assertEquals(mail_message_1.subject, "Test")
+
+    def test_email_subject_template_multi(self):
+        self.template_2 = self.env["mail.message.custom.subject"].create({
+            "name": "Test template 2",
+            "model_id": self.env.ref("account.model_account_invoice").id,
+            "subtype_ids": [(6, 0, [
+                self.env.ref('mail.mt_comment').id,
+            ])],
+            "subject_template": "${object.number or 'n/a'} and something different",
+        })
+        # Send message in invoice
+        mail_message_1 = self.invoice_1.message_post(body='Test', subtype='mail.mt_comment')
+        # Get message and check subject
+        self.assertEquals(mail_message_1.subject, "n/a and something different")
+
+        self.template_3 = self.env["mail.message.custom.subject"].create({
+            "name": "Test template 3",
+            "model_id": self.env.ref("account.model_account_invoice").id,
+            "subtype_ids": [(6, 0, [
+                self.env.ref('mail.mt_comment').id,
+            ])],
+            "subject_template": "${' and yet something else'}",
+            "replace": False,
+        })
+        # Send message in invoice
+        mail_message_2 = self.invoice_1.message_post(body='Test', subtype='mail.mt_comment')
+        # Get message and check subject
+        self.assertEquals(mail_message_2.subject, "n/a and something different and yet something else")
```
Test with partner messages only.

victoralmau reviewed (changes requested):
Need to add .pot files

joao-p-marques commented:
Changes made @victoralmau 

> Need to add .pot files

I think this is added automatically by OCA Bot, right @pedrobaeza?

pedrobaeza commented:
POT is added automatically, but please get used to provide such file as side effect of the most important one: the Spanish translation for new modules, for having them translated since the beginning.

pedrobaeza commented:
Red Travis

joao-p-marques commented:
Ok, thanks for the explanation. I will add it and check travis.

joao-p-marques committed:
[ADD] mail_notification_custom_subject: Apply a custom subject to mail notifications

This module allows you to specify templates to override the subject on the notification
emails sent by Odoo.

TT27191

Co-authored-by: Pedro M. Baeza <pedro.baeza@tecnativa.com>

victoralmau commented on mail_notification_custom_subject/models/mail_message_custom_subject.py:
```diff
@@ -0,0 +1,257 @@
+            res_ids = [res_ids]
+
+        results = dict.fromkeys(res_ids, u"")
+
```
### 
```suggestion
```

victoralmau commented on mail_notification_custom_subject/models/mail_message_custom_subject.py:
```diff
@@ -0,0 +1,257 @@
+        except Exception:
+            _logger.info("Failed to load template %r", template_txt, exc_info=True)
+            return multi_mode and results or results[res_ids[0]]
+
```
```suggestion
```

victoralmau commented on mail_notification_custom_subject/models/mail_message_custom_subject.py:
```diff
@@ -0,0 +1,257 @@
+            if render_result == u"False":
+                render_result = u""
+            results[res_id] = render_result
+
```
```suggestion
```

victoralmau commented on mail_notification_custom_subject/models/mail_message_custom_subject.py:
```diff
@@ -0,0 +1,257 @@
+        if post_process:
+            for res_id, result in results.items():
+                results[res_id] = self.render_post_process(result)
+
```
```suggestion
```

victoralmau commented on mail_notification_custom_subject/models/mail_thread.py:
```diff
@@ -0,0 +1,40 @@
+            ("model_id.model", "=", self._name),
+            ("subtype_ids", "=", subtype_id),
+        ])
+        subject_override = mail_template.render_subject_template(
+            self._name, self.id
+        )
+        subject = subject_override or subject
```
```suggestion
        if mail_template:
            subject_override = mail_template.render_subject_template(
                self._name, self.id
            )
            subject = subject_override or subject
```

joao-p-marques replied:
Thank you! :+1: 
I don't think it would cause problems, though, as the search would return an empty recordset, which [will then be iterated over](https://github.com/OCA/social/pull/650/files#diff-9c0df1d72822413b9e7331e329dd186ca0093c700694cabe007f8def001bf5baR172), right?
Just a curiosity :smile: Change made :+1:

victoralmau commented on mail_notification_custom_subject/tests/test_mail_notification_custom_subject.py:
```diff
@@ -0,0 +1,122 @@
+                "supplier": True,
+            }
+        )
+        # Delete current DB records for testing
+        self.env["mail.message.custom.subject"].search([]).unlink()
```
```suggestion
```

joao-p-marques replied:
I was using this to get correct test results in case there are some records presents in the database beforehand that would produce different results. 
This transaction is then rolled back at the end of the test, as it is a `TransactionCase`, so it wouldn't affect the real DB. 
What is the correct way to do it, then?

pedrobaeza replied:
This is risky indeed. Perform a search before doing `message_post` and after, and then the difference will be the new created message.

joao-p-marques replied:
Thanks. I ended up deleting these lines, as suggested by @victoralmau as I misunderstood the problem :sweat_smile:

victoralmau commented on mail_notification_custom_subject/tests/test_mail_notification_custom_subject.py:
```diff
@@ -0,0 +1,122 @@
+        )
+        # Get message and check subject
+        self.assertEquals(mail_message_1.subject, "Test partner 1 and something more")
+
```
```suggestion
```

victoralmau commented on mail_notification_custom_subject/tests/test_mail_notification_custom_subject.py:
```diff
@@ -0,0 +1,122 @@
+        )
+        # Get message and check subject
+        self.assertEquals(mail_message_2.subject, "Test partner 2 and something more")
+
```
```suggestion
```

victoralmau commented on mail_notification_custom_subject/tests/test_mail_notification_custom_subject.py:
```diff
@@ -0,0 +1,122 @@
+        self.assertEquals(
+            mail_message_1.subject, "Test partner 1 and something different"
+        )
+
```
```suggestion
```

victoralmau commented on mail_notification_custom_subject/views/mail_notification_custom_subject_views.xml:
```diff
@@ -0,0 +1,50 @@
+                </form>
+            </field>
+        </record>
+
```
```suggestion
```

victoralmau commented on mail_notification_custom_subject/views/mail_notification_custom_subject_views.xml:
```diff
@@ -0,0 +1,50 @@
+                </tree>
+            </field>
+        </record>
+
```
```suggestion
```

victoralmau commented on mail_notification_custom_subject/views/mail_notification_custom_subject_views.xml:
```diff
@@ -0,0 +1,50 @@
+            <field name="view_mode">form,tree</field>
+            <field name="view_id" ref="mail_notification_custom_subject_tree" />
+        </record>
+
```
```suggestion
```

victoralmau commented on mail_notification_custom_subject/views/mail_notification_custom_subject_views.xml:
```diff
@@ -0,0 +1,50 @@
+
+        <menuitem id="menu_mail_notification_custom_subject" parent="base.menu_email" action="action_mail_notification_custom_subject_tree_all"
+                  sequence="21"/>
+
```
```suggestion
```

victoralmau reviewed (changes requested):
Please, remove empty lines and other changes.

victoralmau commented on mail_notification_custom_subject/models/mail_message_custom_subject.py:
```diff
@@ -0,0 +1,243 @@
+
+    @api.multi
+    def render_subject_template(self, model, res_id):
+        subject_override = ""
```
```suggestion
        subject_override = False
```

joao-p-marques replied:
Thanks. However, I can't set this to `False` as I later need to append other strings to this. What I can do instead is to change the return statement to:
```python
return subject_override or False
```
WDYT?

joao-p-marques commented:
All changes done @victoralmau

pedrobaeza commented on mail_notification_custom_subject/README.rst:
```diff
@@ -0,0 +1,99 @@
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module allows you to specify templates to override the emails sent by Odoo
+in the chatter.
```
```suggestion
This module allows you to specify templates to override the subject on the notification emails
sent by Odoo.
```

pedrobaeza commented on mail_notification_custom_subject/README.rst:
```diff
@@ -0,0 +1,99 @@
+* Create a new template.
+
+    * The field **Subject Template** accepts `Jinjah<ttps://jinja.palletsprojects.com/en/2.11.x/>` expressions.
+    * The field **Replace** specifies if the template should replace existing templates or append to them.
```
```suggestion
    * The field **Replace** specifies if the template should replace existing content or append to it.
```

pedrobaeza commented on mail_notification_custom_subject/README.rst:
```diff
@@ -0,0 +1,99 @@
+* Create a new template.
+
+    * The field **Subject Template** accepts `Jinjah<ttps://jinja.palletsprojects.com/en/2.11.x/>` expressions.
+    * The field **Replace** specifies if the template should replace existing templates or append to them.
```
Missing model definition

pedrobaeza replied:
Till here should go on CONFIGURE.rst, not USAGE.rst

pedrobaeza commented on mail_notification_custom_subject/models/mail_message_custom_subject.py:
```diff
@@ -0,0 +1,243 @@
+_logger = logging.getLogger(__name__)
+
+
+def format_date(env, date, pattern=False):
```
Why redefining all these methods? Can't you import them?
This is linked with the following comment

pedrobaeza commented on mail_notification_custom_subject/models/mail_message_custom_subject.py:
```diff
@@ -0,0 +1,243 @@
+        return subject_override or False
+
+    @api.multi
+    def _render_template(self, template_txt, model, res_ids, post_process=False):
```
Aren't you able to call existing methods in mail.template or similar to avoid to rewrite all this stuff?

pedrobaeza commented on mail_notification_custom_subject/models/mail_message_custom_subject.py:
```diff
@@ -0,0 +1,243 @@
+
+    @api.multi
+    def render_subject_template(self, model, res_id):
+        subject_override = ""
```
This is not taking into account existing "core" subject. IMO this should be done on `mail.thread`, not here.

pedrobaeza commented on mail_notification_custom_subject/models/mail_thread.py:
```diff
@@ -0,0 +1,41 @@
+                self._name, self.id
+            )
+            subject = subject_override or subject
+        return super(MailThread, self).message_post(
```
```suggestion
        return super().message_post(
```

pedrobaeza commented on mail_notification_custom_subject/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1,3 @@
+* Tecnativa <https://www.tecnativa.com>
+
+    * João Marques
```
Please include me as designer

pedrobaeza commented on mail_notification_custom_subject/static/description/icon.svg:
```diff
@@ -0,0 +1,248 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
This icon is not correct (it's from `mail_debrand`). Adapt it or remove it.

joao-p-marques commented:
Thanks for the suggestions @pedrobaeza!
Changes made

pedrobaeza commented on mail_notification_custom_subject/models/mail_message_custom_subject.py:
```diff
@@ -0,0 +1,40 @@
+    name = fields.Char(string="Template Name")
+    model_id = fields.Many2one(
+        comodel_name="ir.model",
+        string="Models",
```
```suggestion
        string="Model",
```

pedrobaeza commented on mail_notification_custom_subject/README.rst:
```diff
@@ -0,0 +1,106 @@
+* Go to **Settings > Technical > Email > Subject Replacement Templates**
+* Create a new template.
+
+    * The field **Models** specifies the models to which the subject template should apply in the emails sent by Odoo
```
```suggestion
    * The field **Model** specifies the model to which the subject template should apply in the notification emails sent by Odoo.
```

pedrobaeza commented on mail_notification_custom_subject/models/mail_thread.py:
```diff
@@ -0,0 +1,49 @@
+            subtype = subtype or 'mt_note'
+            if '.' not in subtype:
+                subtype = 'mail.%s' % subtype
+            subtype_id = self.env['ir.model.data'].xmlid_to_res_id(subtype)
```
Don't fail if invalid:
```suggestion
            subtype_id = self.env['ir.model.data'].xmlid_to_res_id(subtype, raise_if_not_found=False)
        if subtype_id:
            ...
```

pedrobaeza commented on mail_notification_custom_subject/models/mail_thread.py:
```diff
@@ -0,0 +1,49 @@
+            if '.' not in subtype:
+                subtype = 'mail.%s' % subtype
+            subtype_id = self.env['ir.model.data'].xmlid_to_res_id(subtype)
+        mail_templates = self.env["mail.message.custom.subject"].search([
```
Use proper variable names
```suggestion
        custom_subjects = self.env["mail.message.custom.subject"].search([
```

pedrobaeza commented on mail_notification_custom_subject/models/mail_thread.py:
```diff
@@ -0,0 +1,49 @@
+            ("subtype_ids", "=", subtype_id),
+        ])
+        for template in mail_templates:
+            rendered_subject_template = self.env["mail.template"]._render_template(
+                template_txt=template.subject_template,
+                model=self._name, res_ids=self.id
+            )
```
I'm thinking that we shouldn't interrupt by any chance the notification sending:
```suggestion
            try:
                rendered_subject_template = self.env["mail.template"]._render_template(
                    template_txt=template.subject_template,
                    model=self._name, res_ids=self.id
                )
            except Exception:
                rendered_subject_template = False
```

joao-p-marques replied:
Ok, thanks! I just made the change. 
However, can't that make it harder to debug if a subject template is well constructed or not? The `_render_template` function throws some helpful exceptions for that (https://github.com/odoo/odoo/blob/fd56023d26dcebb1503fe8d4ece6d6125dfc3e90/addons/mail/models/mail_template.py#L338). Should I catch them and do something with them?

pedrobaeza replied:
You would already have the log entry, so no, nothing should block the notification.

joao-p-marques replied:
Ok, nice! :+1:

pedrobaeza commented:
And the rest of the asked changes?

joao-p-marques commented:
Sorry, I missed those earlier. Changes made.
I also added 1 more check bellow to make sure we don't get `subject = None` when trying to append to an existing one.

pedrobaeza commented:
Travis is red. Please also get full coverage.

joao-p-marques commented:
The latest commit should improve test coverage. Everything else is done, I think @pedrobaeza

pedrobaeza commented on mail_notification_custom_subject/models/mail_thread.py:
```diff
@@ -0,0 +1,68 @@
+        mail_auto_delete=True,
+        **kwargs
+    ):
+        new_message = super().message_post(
```
If you call super before, the mail sending is done in this stage AFAIK

joao-p-marques replied:
From my testing the email is only sent after I return the new message
at the end.

joao-p-marques commented:
From my testing the email is only sent after I return the new message 
at the end.

On Wed, Jan 13, 2021 at 07:15, Pedro M. Baeza 
<notifications@github.com> wrote:
> *@pedrobaeza* commented on this pull request.
> 
> In mail_notification_custom_subject/models/mail_thread.py 
> <https://github.com/OCA/social/pull/650#discussion_r556595190>:
> 
> > +    @api.returns("mail.message", lambda value: value.id)
> +    def message_post(
> +        self,
> +        body="",
> +        subject=None,
> +        message_type="notification",
> +        subtype=None,
> +        parent_id=False,
> +        attachments=None,
> +        notif_layout=False,
> +        add_sign=True,
> +        model_description=False,
> +        mail_auto_delete=True,
> +        **kwargs
> +    ):
> +        new_message = super().message_post(
> If you call super before, the mail sending is done in this stage AFAIK
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/OCA/social/pull/650#pullrequestreview-567322703>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AJJMDDQ6FFVYJEK45DRX7CLSZW2JHANCNFSM4VF2JWBQ>.
>

joao-p-marques commented on mail_notification_custom_subject/models/mail_thread.py:
```diff
@@ -0,0 +1,75 @@
+                        template_txt=template.subject_template,
+                        model=self._name,
+                        res_ids=self.id,
+                    )
```
Should we still check that `rendered_subject_template` has something?
```suggestion
                    )
                    if not rendered_subject_template:
                        continue
```

pedrobaeza replied:
No being everything inside try

joao-p-marques replied:
Ok, seems good then :+1:

joao-p-marques reviewed (commented):
Thank you for the changes @pedrobaeza :smiley: 

I just have a small comment/doubt.

pedrobaeza commented:
/ocabot merge nobump
